File: src/main/java/implementslegendkt/mod/vaultjp/screen/view/ButtonViewDSL.java
package implementslegendkt.mod.vaultjp.screen.view;

import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import implementslegendkt.mod.vaultjp.screen.DecentScreen;
import implementslegendkt.mod.vaultjp.screen.View;
import implementslegendkt.mod.vaultjp.screen.ViewInteractor;
import net.minecraft.client.gui.GuiComponent;
import net.minecraft.client.renderer.GameRenderer;
import net.minecraft.client.renderer.Rect2i;
import net.minecraft.resources.ResourceLocation;
import oshi.util.tuples.Pair;

import java.util.ArrayList;
import java.util.function.Function;
import java.util.function.Supplier;

public class ButtonViewDSL implements View {

    private boolean clicked = false;
    public boolean isClicked(){return clicked;}
    public Supplier<ResourceLocation> texture;
    public Supplier<Rect2i> srcRect;
    public Supplier<Pair<Integer,Integer>> atlasSize;
    public Supplier<Pair<Integer,Integer>> pos = ()->new Pair<>(0,0);
    public Runnable onClick = ()->{};

    public Function<Boolean, Boolean> shouldHighlight = (hovering)->hovering;

    public static class ButtonInteractor implements ViewInteractor<ButtonViewDSL> {

        @Override
        public void clear() {
            views.clear();
        }
        private ArrayList<ButtonViewDSL> views = new ArrayList<>();
        @Override
        public void addView(ButtonViewDSL dsl) {
            views.add(dsl);
        }

        @Override
        public <S extends DecentScreen<S,?>> void renderViews(S screen, PoseStack stack, int cursorX, int cursorY) {
            for (var view:views) {
                RenderSystem.setShader(GameRenderer::getPositionTexShader);
                RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
                RenderSystem.setShaderTexture(0, view.texture.get());
                var rect = view.srcRect.get();
                var position = view.pos.get();
                var atlas = view.atlasSize.get();
                screen.blit(stack, position.getA(), position.getB(), screen.getBlitOffset(), rect.getX(), rect.getY(),rect.getWidth(), rect.getHeight(),atlas.getA(),atlas.getB());

                var hoveringX = cursorX>position.getA() && cursorX< position.getA()+rect.getWidth();
                var hoveringY = cursorY>position.getB() && cursorY< position.getB()+rect.getHeight();
                if(view.shouldHighlight.apply(hoveringX && hoveringY)){
                    RenderSystem.disableDepthTest();
                    GuiComponent.fill(stack, position.getA(),position.getB(), position.getA() + rect.getWidth(), position.getB() + rect.getHeight(), 0x40ffffff);
                    RenderSystem.enableDepthTest();
                }
            }
        }

        @Override
        public <S extends DecentScreen<S,?>> void click(S screen, int cursorX, int cursorY,int key) {
            for (var view:views) {
                var position = view.pos.get();
                var srcrect = view.srcRect.get();
                if (cursorX>position.getA() && cursorX< position.getA()+srcrect.getWidth() && cursorY>position.getB() && cursorY< position.getB()+srcrect.getHeight()){
                    //view.clicked=true;
                    view.onClick.run();
                }else{
                    //view.clicked=false;
                }


            }

        }

    }

}

File: src/main/java/implementslegendkt/mod/screenlegends/view/ButtonViewDSL.kt
package implementslegendkt.mod.screenlegends.view

import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.vertex.PoseStack
import implementslegendkt.mod.screenlegends.DecentScreen
import implementslegendkt.mod.screenlegends.View
import implementslegendkt.mod.screenlegends.ViewInteractor
import implementslegendkt.mod.screenlegends.forEnabledViews
import net.minecraft.client.gui.GuiComponent
import net.minecraft.client.renderer.GameRenderer
import net.minecraft.client.renderer.Rect2i
import net.minecraft.resources.ResourceLocation
import java.util.function.Function
import java.util.function.Supplier

class ButtonViewDSL : View {
    var texture: ()->ResourceLocation? = { null }
    var srcRect: ()->Rect2i? = { null }
    var atlasSize: ()->Pair<Int, Int>? = { null }
    var pos: ()->Pair<Int, Int> = { 0 to 0 }
    var onClick: ()->Unit = {}

    var shouldHighlight: Function<Boolean, Boolean> = Function { hovering: Boolean -> hovering }

    override var enabled: () -> Boolean = {true}

    class ButtonInteractor : ViewInteractor<ButtonViewDSL> {
        override fun clear() {
            views.clear()
        }

        override val views = ArrayList<ButtonViewDSL>()
        override fun addView(dsl: ButtonViewDSL) {
            views.add(dsl)
        }

        override fun <S : DecentScreen<*, *>> renderViews(screen: S, stack: PoseStack?, cursorX: Int, cursorY: Int) {
            forEnabledViews{view->
                RenderSystem.setShader { GameRenderer.getPositionTexShader() }
                RenderSystem.setShaderColor(1.0f, 1.0f, 1.0f, 1.0f)
                view.texture()?.let { RenderSystem.setShaderTexture(0, it) }
                val rect = view.srcRect()?:return@forEnabledViews
                val position = view.pos()
                val atlas = view.atlasSize()?:return@forEnabledViews
                GuiComponent.blit(
                    stack,
                    position.first,
                    position.second,
                    screen.getBlitOffset(),
                    rect.x.toFloat(),
                    rect.y.toFloat(),
                    rect.width,
                    rect.height,
                    atlas.first,
                    atlas.second
                )

                val hoveringX = cursorX > position.first && cursorX < position.first + rect.width
                val hoveringY = cursorY > position.second && cursorY < position.second + rect.height
                if (view.shouldHighlight.apply(hoveringX && hoveringY)) {
                    RenderSystem.disableDepthTest()
                    GuiComponent.fill(
                        stack,
                        position.first,
                        position.second,
                        position.first + rect.width,
                        position.second + rect.height,
                        0x40ffffff
                    )
                    RenderSystem.enableDepthTest()
                }
            }
        }

        override fun <S : DecentScreen<*, *>> click(screen: S, cursorX: Int, cursorY: Int, key: Int) {
            forEnabledViews{view->
                val position = view.pos()
                val srcrect = view.srcRect()?:return@forEnabledViews
                if (cursorX > position.first && cursorX < position.first + srcrect.width && cursorY > position.second && cursorY < position.second + srcrect.height) {
                    //view.clicked=true;
                    view.onClick()
                } else {
                    //view.clicked=false;
                }
            }
        }
    }
}
